com.unity.netcode.gameobjects/Runtime/Serialization/BufferSerializer.cs

```diff
@@ -0,0 +1,343 @@
+        /// function instead of SerializeValue. SerializeValue will incur a boxing allocation,
+        /// SerializeNetworkSerializable will not.
+        ///
+        /// A definition of SerializeValue that doesn't allocate can't be created because C#
+        /// doesn't allow overriding generics based solely on the constraint, so this would conflict
+        /// with WriteValue&lt;T&gt;(ref T value) where T: unmanaged
+        /// 
+        /// Throws OverflowException if the end of the buffer has been reached.
+        /// Write buffers will grow up to the maximum allowable message size before throwing OverflowException.
+        /// </summary>
+        /// <param name="value">Value to serialize</param>
+        public void SerializeNetworkSerializable<T>(ref T value) where T : INetworkSerializable
+        {
+            m_Implementation.SerializeNetworkSerializable(ref value);
+        }
+
+        /// <summary>
+        /// Serialize an INetworkSerializable
+        /// 
+        /// Throws OverflowException if the end of the buffer has been reached.
+        /// Write buffers will grow up to the maximum allowable message size before throwing OverflowException.
+        /// </summary>
+        /// <param name="value">Value to serialize</param>
+        public void SerializeValue(ref INetworkSerializable value)
+        {
+            m_Implementation.SerializeValue(ref value);
+        }
+
+        /// <summary>
+        /// Serialize a GameObject
+        /// 
+        /// Throws OverflowException if the end of the buffer has been reached.
+        /// Write buffers will grow up to the maximum allowable message size before throwing OverflowException.
+        /// </summary>
+        /// <param name="value">Value to serialize</param>
+        public void SerializeValue(ref GameObject value)
+        {
+            m_Implementation.SerializeValue(ref value);
+        }
+
+        /// <summary>
+        /// Serialize a NetworkObject
+        /// 
+        /// Throws OverflowException if the end of the buffer has been reached.
+        /// Write buffers will grow up to the maximum allowable message size before throwing OverflowException.
+        /// </summary>
+        /// <param name="value">Value to serialize</param>
+        public void SerializeValue(ref NetworkObject value)
+        {
+            m_Implementation.SerializeValue(ref value);
+        }
+
+        /// <summary>
+        /// Serialize a NetworkBehaviour
+        /// 
+        /// Throws OverflowException if the end of the buffer has been reached.
+        /// Write buffers will grow up to the maximum allowable message size before throwing OverflowException.
+        /// </summary>
+        /// <param name="value">Value to serialize</param>
+        public void SerializeValue(ref NetworkBehaviour value)
+        {
+            m_Implementation.SerializeValue(ref value);
+        }
```

Comment:
I wouldn't re-introduce these (both here and also reader/writer impls) because we know that we want to have `struct NetworkObjectReference` etc. for these scenarios. Killing these here and in reader/writer also kills NetworkManager.SpawnManager deps too — so, an easy win-win?

Reply:
(I'd also add Read/Write Object to this context too, I remember calling that out earlier but I wanted to call out again to stress a little bit more)

Reply:
I'll say the same thing I said before.. I want to finish INetworkMessage implementation before I kill these. If it turns out they're unused/unneeded once everything's done, I'll happily kill them, but I don't want to kill them now and have to resurrect them later. Converting NetworkObject and NetworkBehaviour to be INetworkSerializable should be trivial... GameObject not so much, but we can also just not support game objects and demand that the player send us the NetworkObject.